### PR TITLE
Add threading when writing events to elasticsearch

### PR DESCRIPTION
**Pull request by rccrdpccl**
This PR is trying to accomplish the following:
* Add concurrency to the process of writing cluster events to elasticsearch.
  * this has been done with the means of threading through a shared queue
* Simplify code
* * Code was in mostly one file, we tried to split in more files so that future work can be easily done. This is far from a perfect refactor, but the aim is to simplify future work and improve iterating over this

Although the aim is limited to reducing overall time for parsing events, all type of suggestions are welcomed (even structural).

The product of this PR has been tested in integration, and it starts with no obvious errors.

### CODE REVIEW

File: assisted-events-scrape/storage/cluster_events_storage.py
Comment (rccrdpccl) [inline assisted-events-scrape/storage/cluster_events_storage.py:1]: We've moved all related code to "Process and store events in elastic" to this module.
It's far from ideal has this hasn't a single responsibility, and we'll definitely tackle this in the future. For now we can consider this a "Laundry drawer" for "elastic-related-stuff".

File: assisted-events-scrape/repositories/event_repository.py
Comment (rccrdpccl) [inline assisted-events-scrape/repositories/event_repository.py:1]: We have moved here all code related to retrieving events.
This should be changed in the future, but for now is as it was before.

File: assisted-events-scrape/workers/cluster_events_worker.py
Comment (eliorerz) [inline assisted-events-scrape/workers/cluster_events_worker.py:34]: Unresolved attribute reference 'queue' for class 'ClusterEventsWorker'

File: assisted-events-scrape/utils/counters.py
Comment (eliorerz) [inline assisted-events-scrape/utils/counters.py:15]: Type doesn't match - Expected type 'None', got 'int' instead 

```suggestion
    def get_errors(self) -> union[int, None]:
```

File: assisted-events-scrape/utils/counters.py
Comment (eliorerz) [inline assisted-events-scrape/utils/counters.py:39]: timedelta gets float but you gave it as an argument another timedelta

Comment (eliorerz) [reply assisted-events-scrape/utils/counters.py:39]: On other places you are passing mostly int, so you might want to change the type hint annotations

File: assisted-events-scrape/events_scrape/events_scrape.py
Comment (eliorerz) [inline assisted-events-scrape/events_scrape/events_scrape.py:76]: This function is way too long for init. Please try to split it into functions.

File: assisted-events-scrape/events_scrape/events_scrape.py
Comment (eliorerz) [inline assisted-events-scrape/events_scrape/events_scrape.py:41]: max line length is set to 120

```suggestion
        self.client = ClientFactory.create_client(url=config["inventory_url"], offline_token=config["offline_token"])
```

File: assisted-events-scrape/events_scrape/events_scrape.py
Comment (eliorerz) [inline assisted-events-scrape/events_scrape/events_scrape.py:160]: I think it is better to create this config as dataclass (autocomplete, type annotations and more)

File: assisted-events-scrape/events_scrape/events_scrape.py
Comment (eliorerz) [inline assisted-events-scrape/events_scrape/events_scrape.py:79]: Unresolved attribute reference 'changes' for class 'ScrapeEvents'

File: assisted-events-scrape/events_scrape/events_scrape.py
Comment (eliorerz) [inline assisted-events-scrape/events_scrape/events_scrape.py:82]: Unresolved attribute reference 'error_counter' for class 'ScrapeEvents'

File: assisted-events-scrape/events_scrape/events_scrape.py
Comment (eliorerz) [inline assisted-events-scrape/events_scrape/events_scrape.py:64]: Also here, I think it is better to create a dataclass configuration object instead of dict

File: assisted-events-scrape/events_scrape/events_scrape.py
Comment (eliorerz) [inline assisted-events-scrape/events_scrape/events_scrape.py:18]: Add the objects to the `workeres.__init__` file so you will be able to import the object from the package and not from each module
e.g. 
```suggestion
from workers import ClusterEventsWorker
```

File: assisted-events-scrape/events_scrape/events_scrape.py
Comment (eliorerz) [inline assisted-events-scrape/events_scrape/events_scrape.py:16]: Add all the Repository objects to the __init__ file 


```suggestion
from repositories import ClusterRepository, EventRepository
```

File: requirements.txt
Comment (osherdp) [inline requirements.txt:12]: nose is probably discontinued, so it will be a good idea to move to pytest:
https://nose.readthedocs.io/en/latest/#note-to-users

File: assisted-events-scrape/tests/test_cluster_events_worker.py
Comment (eliorerz) [inline assisted-events-scrape/tests/test_cluster_events_worker.py:1]: Why unittest? We are using pytest on test-infra and I think it is better to use the same infrastructure for testing on all of our projects.

File: assisted-events-scrape/repositories/event_repository.py
Comment (eliorerz) [inline assisted-events-scrape/repositories/event_repository.py:13]: This line is not needed  - `Local variable 'event_list' value is not used `

File: Makefile
Comment (osherdp) [inline Makefile:35]: Not a blocker for this PR, but we should add this as a prow job

**Review by osherdp — CHANGES_REQUESTED** [review]

File: assisted-events-scrape/events_scrape/events_scrape.py
Comment (eliorerz) [inline assisted-events-scrape/events_scrape/events_scrape.py:142]: Function name should be lowercase

**Review by eliorerz — CHANGES_REQUESTED** [review]
Partially reviewed

File: assisted-events-scrape/tests/test_cluster_events_worker.py
Comment (rccrdpccl) [reply assisted-events-scrape/tests/test_cluster_events_worker.py:1]: No real reason for unittest, just did not think about checking out test-infra. I agree with you, I will change it to pytest.

Comment (rccrdpccl) [reply assisted-events-scrape/tests/test_cluster_events_worker.py]: @eliorerz what package would you recommend for mocking? AFAIU unittest.mock is the one recommended also when using pytest, is this correct?

Comment (osherdp) [reply assisted-events-scrape/tests/test_cluster_events_worker.py]: yes, I think ``unittest.mock`` is still the way to go

File: Makefile
Comment (rccrdpccl) [reply Makefile]: I'd love to, how can this be done? Do you have a link to any similar PR?

Comment (osherdp) [reply Makefile]: this one is a good example: https://github.com/openshift/release/pull/21127/files

Comment (rccrdpccl) [reply Makefile]: OK, AFAIU this won't work until this is merged (it will work only for this PR).
I will follow up with CI integration once this gets merged

Comment (osherdp) [reply Makefile]: sure, no problem
that's a good plan

File: assisted-events-scrape/events_scrape/events_scrape.py
Comment (michaellevy101) [inline assisted-events-scrape/events_scrape/events_scrape.py:98]: when res would be empty?

File: assisted-events-scrape/events_scrape/events_scrape.py
Comment (michaellevy101) [reply assisted-events-scrape/events_scrape/events_scrape.py:155]: agree

File: assisted-events-scrape/events_scrape/events_scrape.py
Comment (rccrdpccl) [reply assisted-events-scrape/events_scrape/events_scrape.py:122]: As it stands the bare minimum is main process + spawned thread

File: assisted-events-scrape/repositories/cluster_repository.py
Comment (michaellevy101) [inline assisted-events-scrape/repositories/cluster_repository.py:16]: please log even for "Not found". "the cluster has been deleted" or similar message

File: assisted-events-scrape/storage/cluster_events_storage.py
Comment (michaellevy101) [inline assisted-events-scrape/storage/cluster_events_storage.py:23]: for now it's ok, but the storage shouldn't contain the AI client

File: assisted-events-scrape/repositories/cluster_repository.py
Comment (rccrdpccl) [reply assisted-events-scrape/repositories/cluster_repository.py:16]: This will be logged (and sent to sentry) by https://github.com/openshift-assisted/assisted-events-scrape/pull/49/files#diff-9bf5d2c8ddd304179959d273e721b6e244a544883b8139c41abb607828378018R47

Not sure is worth adding further info to this specific message? WDYT?

File: assisted-events-scrape/events_scrape/events_scrape.py
Comment (michaellevy101) [reply assisted-events-scrape/events_scrape/events_scrape.py:98]: if you will change the configuration to be a data class it won't be necessary. right?

File: assisted-events-scrape/events_scrape/events_scrape.py
Comment (michaellevy101) [reply assisted-events-scrape/events_scrape/events_scrape.py:122]: that's ok. please document it

File: assisted-events-scrape/repositories/cluster_repository.py
Comment (michaellevy101) [reply assisted-events-scrape/repositories/cluster_repository.py:16]: we catch this exception

File: assisted-events-scrape/storage/cluster_events_storage.py
Comment (rccrdpccl) [reply assisted-events-scrape/storage/cluster_events_storage.py:23]: I agree

File: assisted-events-scrape/events_scrape/events_scrape.py
Comment (tsorya) [inline assisted-events-scrape/events_scrape/events_scrape.py:60]: why not to use thread pool?

File: assisted-events-scrape/events_scrape/events_scrape.py
Comment (tsorya) [inline assisted-events-scrape/events_scrape/events_scrape.py:79]: sleep here is a little bit strange for me as there is not loop and i needed to go to the calling function to understand why it was needed

Comment (rccrdpccl) [reply assisted-events-scrape/events_scrape/events_scrape.py:79]: Totally agree, during the next iteration we'll try to organize better this part as well.

Comment (rccrdpccl) [reply assisted-events-scrape/events_scrape/events_scrape.py]: Actually I'll just change it, it only takes 1 minute

File: assisted-events-scrape/tests/test_utils.py
Comment (eliorerz) [inline assisted-events-scrape/tests/test_utils.py:6]: Remove redundant parentheses

File: assisted-events-scrape/tests/test_cluster_repository.py
Comment (eliorerz) [inline assisted-events-scrape/tests/test_cluster_repository.py:8]: Remove redundant parentheses

File: assisted-events-scrape/tests/test_cluster_events_worker.py
Comment (eliorerz) [inline assisted-events-scrape/tests/test_cluster_events_worker.py:8]: Remove redundant parentheses

File: assisted-events-scrape/events_scrape/events_scrape.py
Comment (eliorerz) [inline assisted-events-scrape/events_scrape/events_scrape.py:52]: Instance attribute errors_before_restart defined outside __init__

File: assisted-events-scrape/workers/cluster_events_worker.py
Comment (eliorerz) [inline assisted-events-scrape/workers/cluster_events_worker.py:17]: Remove redundant parentheses

File: assisted-events-scrape/events_scrape/__main__.py
Comment (eliorerz) [inline assisted-events-scrape/events_scrape/__main__.py:5]: `Cannot find reference 'main' in '__init__.py'`
Why did you removed the relative import?

File: assisted-events-scrape/repositories/__init__.py
Comment (eliorerz) [inline assisted-events-scrape/repositories/__init__.py:2]: Need to declare `__all__` for defining what symbols in a module will be exported

File: assisted-events-scrape/storage/__init__.py
Comment (eliorerz) [inline assisted-events-scrape/storage/__init__.py:1]: Need to declare `__all__` for defining what symbols in a module will be exported

File: assisted-events-scrape/utils/__init__.py
Comment (eliorerz) [inline assisted-events-scrape/utils/__init__.py:2]: Need to declare `__all__` for defining what symbols in a module will be exported

File: assisted-events-scrape/workers/__init__.py
Comment (eliorerz) [inline assisted-events-scrape/workers/__init__.py:1]: Need to declare `__all__` for defining what symbols in a module will be exported

**Review by eliorerz — CHANGES_REQUESTED** [review]
It looks great! I added few minor comments.
Also note that it's is better to set the class properties as protected and not public (leading underscore)

File: assisted-events-scrape/events_scrape/events_scrape.py
Comment (rccrdpccl) [inline assisted-events-scrape/events_scrape/events_scrape.py:27]: Init is quite big now, but we'll refactor all this in the near future

**Comment by rccrdpccl** [comment]
/hold

holding it to be present when it gets deployed on stage with CD

**Pull request merged**
